Major PR for Sonos Playlist Management.

dundeemt opened the pull request:
Add base method reorder_sonos_playlist and a number of helpers, documentation and tests.

dundeemt commented:
blocking on issue #353, I have a commit to fix the flake8 errors but I can't run my tests since .get_sonos_playlists() is now failing.

Other: my local flake8 is passing without issue, it fails on the CI.  I've got corrections but am waiting to commit until I can get my tests to start running again.

lawrenceakka commented:
I have fixed #353 - sorry about that. Sonos can't make its mind up about whether a playlistContainer is a `<container>` or an `<item>`. I re-ran the Travis job, but you are still failing lint checking :disappointed:

dundeemt commented:
@lawrenceakka no worries.  Working with Sonos can be interesting.  I need to go through the CI script and figure out what the difference is between flake8 on Travis and me running it locally.  The mismatched closed was not showing locally when I ran pep8 and flake8 against my local source.  

Yeah, soco/compat.py -- solves my `unicode` problem.

Is the pylint check new or changed?  I haven't made any changes to data_structures.py in the PR.

dundeemt commented:
Looked at data_structures.DidlObject.from_element, the source of the pylint R0914 too-many-locals error and I am not sure what can be reasonably done outside of telling pylint to ignore it for that method.  The code looks tight and readable.

lawrenceakka commented:
@dundeemt I have no idea why pylint has suddenly started complaining. I have just run up against this too.  Stick a `#pylint: disable=too-many-locals` at an appropriate place.

Your patch looks like it will be very useful.  I will provide some inline comments when I get a moment. One thing I am not so sure about is the syntax for specifying multiple tracks. I would much prefer to use something more like Python's slice syntax, instead of inventing something new. You will see that I have just put in a PR (#354) to change this for ListofMusicItems (and hence the Queue class), so that they can support multiple slice indices (e.g. `playlist[1:3, 6, 20:12:-1]` will return a list of the second and third, fifth, and 20th to 13th elements of the playlist. 

Lawrence

lawrenceakka commented on soco/core.py:
```diff
@@ -134,6 +135,7 @@ class SoCo(_SocoSingletonBase):
         get_composers -- Get composers from the music library
         get_tracks -- Get tracks from the music library
         get_playlists -- Get playlists from the music library
+        get_sonos_playlists -- Get Sonos playlists from the music library.
```
It is not getting playlists from the music library (which, strictly, is the locally stored music, i.e. not Spotify etc).

dundeemt replied:
It wasn't originally listed, so I added it to the main docstring.  I've
modified it now say:
  get_sonos_playlists -- Get Sonos playlists.

Too little or just right?

On Mon, Sep 7, 2015 at 10:44 AM, Lawrence Akka <notifications@github.com>
wrote:

> In soco/core.py
> <https://github.com/SoCo/SoCo/pull/352#discussion_r38870472>:
>
> > @@ -134,6 +135,7 @@ class SoCo(_SocoSingletonBase):
> >          get_composers -- Get composers from the music library
> >          get_tracks -- Get tracks from the music library
> >          get_playlists -- Get playlists from the music library
> > +        get_sonos_playlists -- Get Sonos playlists from the music library.
>
> It is not getting playlists from the music library (which, strictly, is
> the locally stored music, i.e. not Spotify etc).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/352/files#r38870472>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473

lawrenceakka commented on soco/core.py:
```diff
@@ -1255,6 +1266,34 @@ def get_sonos_playlists(self, *args, **kwargs):
+        matches the attribute specified.
+
+        Args:
+            attr_name (str): DidlPlaylistContainer attribute to compare.
```
"e.g. title, or ..." would be useful here

dundeemt replied:
I've modified the string to:
            attr_name (str): DidlPlaylistContainer attribute to compare.
The
                most useful being: 'title' and 'item_id'.


On Mon, Sep 7, 2015 at 10:46 AM, Lawrence Akka <notifications@github.com>
wrote:

> In soco/core.py
> <https://github.com/SoCo/SoCo/pull/352#discussion_r38870587>:
>
> > @@ -1255,6 +1266,34 @@ def get_sonos_playlists(self, *args, **kwargs):
> >          args = tuple(['sonos_playlists'] + list(args))
> >          return self.get_music_library_information(*args, **kwargs)
> >
> > +    def get_sonos_playlist_by_attr(self, attr_name, match):
> > +        """ Return the first Sonos Playlist DidlPlaylistContainer that
> > +        matches the attribute specified.
> > +
> > +        Args:
> > +            attr_name (str): DidlPlaylistContainer attribute to compare.
>
> "e.g. title, or ..." would be useful here
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/352/files#r38870587>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473

lawrenceakka commented on soco/core.py:
```diff
@@ -1255,6 +1266,34 @@ def get_sonos_playlists(self, *args, **kwargs):
+            match (str): Value to match.
+
+        Returns:
+            (:py:class:`~.soco.data_structures.DidlPlaylistContainer`): The
```
I don't think you need the `:py:` here since python is the default in Sphinx. Do you need the "~"? I haven't checked

dundeemt replied:
Just using (DidlPlaylistContainer) resulted in no hyperlink.  So I just cut an pasted from another docstring. However, only newer methods are using Napoleon so it most likely came from a different style docstring.  I'll try and remove and see what it does.

KennethNielsen replied:
I don't think this changes with Napoleop. Napoleon is mostly about formatting the list of Args, Returns, and Raises and their type.

dundeemt commented:
This PR has a number of areas where I expect to get feedback from the group.  However, it covers an area currently under-served by soco, Sonos Playlist management.  

The underlying service is complex and not well documented and i did my best to tame it but still retain the ability to use it optimally.  I've written simplified helper functions for the 3 most common operations, remove a track, move a track and clear the Sonos playlist.  

Areas of note:
 * Indexing - I've used 0-based indexing.  The underlying services refer to the first item in a Sonos Playlist as 0, Python lists use 0 based indexing.  This is in line with the other 'queue' management methods.

* update_id - I am still unsure just how important it is, but the existing add_item_to_sonos_playlist methods deemed it important.  I allow the calling code to set it so a lookup can be avoided, the input parameter is optional and defaults to 0.  I also return it as part of every Sonos playlist mgmt method so the caller can pass it on to the next call if a multi operation is being done.

* object_id - allows for a simple string with the item_id but also accepts an object that has an item_id attribute. I wanted to make it simple for the caller to use what they had available.

* tracks/new_pos 
  * reorder_sonos_playlists - takes a list of ints, a list of int-like or a string containing a comma separated list of numbers.  The tracks/new_pos parameters are expected to be of the same logical type.
  * helper functions accept simple int or int-likes.  (i.e. 1 vs '1')

lawrenceakka commented on soco/core.py:
```diff
@@ -1255,6 +1266,34 @@ def get_sonos_playlists(self, *args, **kwargs):
+
+        Example::
+
+            obj.get_sonos_playlist_by_attr('item_id', 'SQ:3')
```
Please use `device` instead of `obj`.  We will eventually standardise this, but you have to start somewhere!

KennethNielsen replied:
I don't know if I wouldn't prefer an title example instead. Most of the time, users should not have to meddle with these IDs themselves.

dundeemt replied:
are you going to differentiate between any sonos device and a controller/master?  There is a bit of a taxonomy problem. Because a device can be, a stand alone speaker, a pair, a member of a playbar group, etc.  These methods are only valid against a "master".  I need to add `@only_on_master` on these like I did for the `remove_sonos_playlist` PR that is waiting.

KennethNielsen replied:
@dundeemt I think as @lawrenceakka said ``device`` will be the term we will prefer. The fact that you can only do all of these actions on the master, should eventually turn out to be implicit.

lawrenceakka commented on soco/core.py:
```diff
@@ -1833,6 +1872,181 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+        :py:meth:`clear_sonos_playlist`, :py:meth:`move_in_sonos_playlist`,
+        :py:meth:`remove_from_sonos_playlist` for simplified usage.
+
+        update_id - If you have a series of operations, tracking the update_id
```
Please can you explain more - if not in the doctstring then in a normal comment

dundeemt replied:
I've expanded the doc string some what.  But I'll explain in more detail
here and based on feedback, I'll remodify the docstring.

The call can reorder and delete tracks.  In general the underlying call
expects two string arguments, tracks which indicate the 'current' track
position you want to modify and the new_pos that dictate what should happen
to it.  You can, in a single request, specify moving and deleting, but they
happen grouped into two calls.  First doing all the delete requests in an
op, and then doing the moves (however that is with respect to the new track
positions after the deletes.  If you delete a track, all following tracks
are immediately renumbered to fill in the gap. So any move requests need to
take this into account.  It is simpler to think about if you make the
changes individually.

So the first option of list of ints.   tracks=[1,3], new_pos=[0, None]
(since it is using ints, I chose to indicate a delete with None) The
results of this operation would move the second track into the first
position, what was in the 1 position moves to the second postion and the
4th track is deleted from the playlist.

trackA       pos:0
trackB       pos:1
trackC       pos:2
trackD       pos:3

is transformed into:
trackB      pos:0
trackA      pos:1
trackC      pos:2

When the inputs are lists, each pair zip(tracks, new_pos) creates an
underlying call to avTransport.ReorderTracksInSavedQueue.

The same call using a list of int-likes  would look like:  tracks=['1',
'3'], new_pos['0', ''], which would also create two underlying calls.
(note: when using int-likes a delete is specified by an empty string.)

Now, if you really wanted to have fun, you can call reorder with strings,
tracks='1,3,5', new_pos='0,2,4' which would move the 2nd, 4th and 6th track
into the first, third and 5th position in a single underlying call. Using a
string, says: "I know what I'm doing, get out of my way."  It seems to
happen "atomically"  in that the moves happen before the track order
shifts.  However, getting this right can be frustrating when you have
deletes going on at the same time.   Using the helper functions clear, move
and remove make it simpler to think about because the list is being
transformed one step at a time and you only need to supply simple parameters

maybe we should prepend an underscore to reorder_sonos_playlist ->
_reorder_sonos_playlist. To encourage people to use just the helpers. ??

On Mon, Sep 7, 2015 at 10:48 AM, Lawrence Akka <notifications@github.com>
wrote:

> In soco/core.py
> <https://github.com/SoCo/SoCo/pull/352#discussion_r38870720>:
>
> > @@ -1833,6 +1872,181 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
> >              ('AddAtIndex', 4294967295)
> >          ])
> >
> > +    def reorder_sonos_playlist(self, object_id, tracks, new_pos, update_id=0):
> > +        """ Reorder and/or Remove tracks in a Sonos playlist. The underlying
> > +        call is quite complex. See the helper methods
> > +        :py:meth:`clear_sonos_playlist`, :py:meth:`move_in_sonos_playlist`,
> > +        :py:meth:`remove_from_sonos_playlist` for simplified usage.
> > +
> > +        update_id - If you have a series of operations, tracking the update_id
>
> Please can you explain more - if not in the doctstring then in a normal
> comment
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/352/files#r38870720>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473

lawrenceakka commented on soco/core.py:
```diff
@@ -1833,6 +1872,181 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+                a list of int like things. i.e. ``['0', '1',]`` OR it may be a
+                str of comma separated int like things. ``"0,1"``.  Tracks are
+                **0**-based. Meaning the first track is track 0, just like
+                indexing into a Python list.
```
See my comments about multiple slice lists.  Can we do something more pythonic, e.g. `[1, 4:8, 9]` - see also #354

dundeemt replied:
That might be possible but I would like to get capabilities that put sonos_playlists on par with queues first.

lawrenceakka replied:
Agreed.  I think it is too ambitious to go for slicing at the moment.  I amy work on it though, if I get bored .... :wink:

dundeemt commented:
@lawrenceakka  re #354 A read-only structure of course does not make sense for a sonos playlist, it has to be modifiable, after all that what this PR is all about.  I do get your point though about more Pythonic. In general it would be an object that maintained it's state with the backend.  In that regards, it is closer to an ORM than just a subclass of a standard python object (list, queue, ..).  For this PR I was looking to bring parity to Sonos playlists with the current queue methods  ....

KennethNielsen commented:
Hallo @dundeemt 

This looks really good and definitely something we have been missing. I have a few comments about style and preferred arguments and such. I will comment in-line.

KennethNielsen commented on soco/core.py:
```diff
@@ -1255,6 +1266,35 @@ def get_sonos_playlists(self, *args, **kwargs):
+            obj.get_sonos_playlist_by_attr('title', 'Foo')
+
+        """
+        for spl in self.get_sonos_playlists():
```
In general, please write variable names out, especially if we are not short on space. attr for attribute is fine, because it is an established shorthand, but ``spl`` really should be ``sonos_playlist``. "Readability counts" ;)

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
             ('AddAtIndex', 4294967295)
         ])
 
+    def reorder_sonos_playlist(self, object_id, tracks, new_pos, update_id=0):
```
Not only here, but in all methods. I like that you have made the argument that specifies the Sonos playlist flexible, so that it takes both an ID and the playlist object, but please turn the logic around, so that the playlist object is the first class citizen and the ID the alternative, both in the argument name and in the phrasing in the docstring. As I mentioned in one of the other PRs, I think one of the goals of the SoCo API should be to elegantly mask some of all the gory details of Sonos by clear Pythonic types and concepts and I really feel that this falls in that category.

I will make smaller comments with the keyword playlist_object_first in the places were I think there should be a modification.

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+
+        See the helper methods
+        :py:meth:`clear_sonos_playlist`, :py:meth:`move_in_sonos_playlist`,
+        :py:meth:`remove_from_sonos_playlist` for simplified usage.
```
Nice. I like this. Make the advanced methods available and with clear description along with the helpers :)

lawrenceakka replied:
I would emphasise this even more. I think we need a Sphinx "Warning" that using this method, it is possible to mess up your playlist, if you are not very clear about the counting. You should only use this if you know what you are doing, and you probably want to use the helper methods instead!

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+        Examples:
+          To reorder the first two tracks::
+
+            # object_id specified by the item_id
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+            # object_id specified by the item_id
+            obj.reorder_sonos_playlist('SQ:0', tracks=[0, ], new_pos=[1, ])
+            # OR object_id specified by the DidlPlaylistContainer object
+            spl = obj.get_sonos_playlists()[0]
```
please wirte out spl

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+
+          To delete the second track::
+
+            # tracks/new_pos are a list of int
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+            obj.reorder_sonos_playlist('SQ:0', tracks='1', new_pos='')
+
+        Args:
+            object_id (str): The item_id of the Sonos playlist or the Sonos
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+            # OR tracks/new_pos are a list of int-like
+            obj.reorder_sonos_playlist('SQ:0', tracks=['1', ], new_pos=['', ])
+            # OR tracks/new_pos are strings - no transform is done
+            obj.reorder_sonos_playlist('SQ:0', tracks='1', new_pos='')
```
I'm really not sure whether exposing this is more confusing than it is worth. O.k. so this is for performing a single operation. This is already doable with the other methods, so what it saves really is only typing the lists around the single arguments. The downside is that with the ``no transform is done`` comment, you are implying and expectation of knowledge about the underlying protocol. I think, if we are to keep it, scratch the no transform comment.

On another note, I would suggest to make all these multiple track examples, some like to code by example, and it is always easier to infer the singular than the plural

dundeemt replied:
Using this method is the best for removing a bunch of tracks.  It is used by clear_sonos_playlist to remove all tracks in a single operation.  If you had N tracks, without this capability, you would make N network calls.  Using this no-transform method there is 1 call.   A very substantial speed up. This call would have `tracks='1,2,...,n'` and `new_pos=''`

KennethNielsen replied:
Yes, but as I mentioned somewhere later, the reorder_sonos_playlist could (and maybe should) simply turn the iterable inputs into strings form, and always perform just one request.

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+                (:py:class:`~.soco.data_structures.DidlPlaylistContainer`).
+            tracks: (list): list of track indices(int) to reorder. May also be
+                a list of int like things. i.e. ``['0', '1',]`` OR it may be a
+                str of comma separated int like things. ``"0,1"``.  Tracks are
```
as above, suggest to remove the string input form

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+                update_id.
+
+        Raises:
+            SoCoUPnPException: If object_id does not exist.
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+            position_list = [new_pos, ]
+        elif type(tracks) is int:
+            track_list = [tracks, ]
+            position_list = [new_pos or '', ]
```
Is this correct, remember 0 also evaluates to false:
```python
>>> 0 or ''
''
```
that would mean that you cannot use the non-iterable argument form to move a track to the first position.

dundeemt replied:
bug

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+            update_id = response['UpdateID']
+        change = 0
+
+        for track, position in zip(track_list, position_list):
```
Here we need to careful. When I wrote above, that I was unsure if you handled the non-iterable case correctly, it was because I was unsure if the under lying call takes a single position or a list of position. I now see that it takes a list, which of course means that the non-iterable input case is correct.

However, if it does take a list, I would highly suggest to turn all iterable input args into a list and execute only one request. The reason for this is two-fold:

 1. Performance, we should of course never prematurely optimize, but I think this just makes sense. For apps doing network IO, that will always be the "slow thing" and since it is made so easy to do all at once we might as well do it.
 2. Deletions. The significantly more important issue, is that if we do only do one request, we can hope that everything will also be executed as one query on the other end and this is important for deletions. We went through this somewhere else, where the call only took one argument, maybe it was for queue manipulation. Consider this, if I ask this method to delete tracks 0 and 3. After having deleted track 0, the second track that I wanted to delete is now in position 2 **not** 3. This can be fixed by sorting deletions descending, but here, at least of it works, it would be soooo much easier to just do one request. It may also be, that they actually iterate on the other end, in which case, we have to handle this problem here.

Deletions stuff, and the example above, should be tested.

The one thing that I do not know is, if the list form accepts mixing operations, move and delete, or if it might separate them out on the other end. This will make everything more complicated.

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+
+        return change, length, update_id
+
+    def clear_sonos_playlist(self, object_id, update_id=0):
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+        This is a convenience function for :py:meth:`reorder_sonos_playlist`.
+
+        Example:
+            ``obj.clear_sonos_playlist('SQ:0')``
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+
+    def clear_sonos_playlist(self, object_id, update_id=0):
+        """ Clear all tracks from a Sonos playlist.
+        This is a convenience function for :py:meth:`reorder_sonos_playlist`.
```
strictly speaking "convenience method"

lawrenceakka replied:
No space before docstr. please.

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+            ``obj.clear_sonos_playlist('SQ:0')``
+
+        Args:
+            object_id (str): The item_id of the Sonos playlist or
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+        Raises:
+            SoCoUPnPException: See :py:meth:`reorder_sonos_playlist`
+        """
+        sonos_pl = None
```
``sons_pl`` and ``spl`` can be written out more, plenty of space

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+                             ' item_id'.format(object_id))
+        num_tracks = self.browse(ml_item=sonos_pl).total_matches
+        tracks = ','.join([str(x) for x in range(num_tracks)])
+        return self.reorder_sonos_playlist(sonos_pl, tracks=tracks,
```
O.k. I see here that you use the non-iterable input form, but it would only be one extra line not to, if it were to be removed. If this works however, this does seem to indicate that one request is handled as one *query* on the other end, otherwise, this would not work, it would run into non-existing positions along th way.

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+        return self.reorder_sonos_playlist(sonos_pl, tracks=tracks,
+                                           new_pos='', update_id=update_id)
+
+    def move_in_sonos_playlist(self, object_id, track, new_pos, update_id=0):
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+
+        Example::
+
+            obj.move_in_sonos_playlist('SQ:0', track=0, new_pos=1)
```
as above for Lwarence comment about ``device``

also playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+            obj.move_in_sonos_playlist('SQ:0', track=0, new_pos=1)
+
+        Args:
+            object_id (str): The item_id of the Sonos playlist or
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+        Raises:
+            SoCoUPnPException: See :py:meth:`reorder_sonos_playlist`
+        """
+        return self.reorder_sonos_playlist(object_id, int(track),
```
Here you actually require ints and arguments, but you also type convert.

dundeemt replied:
Yup, it's a helper.  It meant to hand hold.  The definition also specifies an int.

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+        return self.reorder_sonos_playlist(object_id, int(track),
+                                           int(new_pos), update_id)
+
+    def remove_from_sonos_playlist(self, object_id, track, update_id=0):
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+
+        Example::
+
+            obj.remove_from_sonos_playlist('SQ:0', track=0)
```
device and playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+            obj.remove_from_sonos_playlist('SQ:0', track=0)
+
+        Args:
+            object_id (str): The item_id of the Sonos playlist or
```
playlist_object_first

KennethNielsen commented on soco/core.py:
```diff
@@ -1833,6 +1873,185 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
+        Raises:
+            SoCoUPnPException: See :py:meth:`reorder_sonos_playlist`
+        """
+        return self.reorder_sonos_playlist(object_id, int(track), None,
```
as above about requiring ints and type converting

dundeemt replied:
also a helper, with well defined track. (int)

KennethNielsen commented on tests/test_integration.py:
```diff
@@ -378,3 +378,362 @@ def test(self, soco):
+class TestReorderSonosPlaylist(object):
+    """ Integration tests for Sonos Playlist Management. """
+    existing_playlists = None
+    pl_name = 'zSocoTestPlayList42'
```
also for test code about writing variable names out, I will not comment any further on this

KennethNielsen commented on tests/test_integration.py:
```diff
@@ -378,3 +378,362 @@ def test(self, soco):
+            self.existing_playlists = soco.get_sonos_playlists()
+        if self.pl_name in [x.title for x in self.existing_playlists]:
+            msg = '%s is an existing playlist.' % self.pl_name
+            pytest.fail(msg)
```
If this is meant for the unlikely case that someone might have a playlist by this name, then you could as python for a UUID. If it is just meant to check for missing cleanup after a failed test, then maybe the error message should say so.

KennethNielsen commented on tests/test_integration.py:
```diff
@@ -378,3 +378,362 @@ def test(self, soco):
+        que = soco.get_queue()
+        test_pl = self._get_test_pl(soco)
+        spl = soco.browse(ml_item=test_pl)
+        tracks = list(reversed(range(len(spl))))
```
Ahhh, I see that you were aware of the deletion problem

KennethNielsen commented on tests/test_integration.py:
```diff
@@ -378,3 +378,362 @@ def test(self, soco):
+        tracks = range(num_tracks)
+        new_pos = list(reversed(tracks))
+
+        args = {'object_id': test_pl.item_id,
```
or maybe just test_pl ;)

KennethNielsen commented on tests/test_integration.py:
```diff
@@ -378,3 +378,362 @@ def test(self, soco):
+        assert count == num_tracks
+        assert update_id != 0
+        spl = soco.browse(ml_item=test_pl)
+        for s_item, q_item in zip(spl, reversed(soco.get_queue())):
```
Testing up against queue here, means that you have to know that the playlist is reset to something that comes from the queue, what about just testing up against test_pl?

KennethNielsen commented on tests/test_integration.py:
```diff
@@ -378,3 +378,362 @@ def test(self, soco):
+            assert s_item.resources[0].uri != q_item.resources[0].uri
+
+    def test_swap_first_two_items(self, soco):
+        """ Test a use case in doc string. Swapping the positions of the first
```
Test a usecase in docstring ???

KennethNielsen commented on tests/test_integration.py:
```diff
@@ -378,3 +378,362 @@ def test(self, soco):
+        """ Test removing first track from Sonos Playlist. """
+        test_pl = self._reset_spl_contents(soco)
+        num_tracks = soco.browse(ml_item=test_pl).total_matches
+        tracks = [0] + range(num_tracks - 1)    # [0, 0, 1, ..., n-1]
```
Does this fail if you go all the way up to num_tracks? Just curious as this might shed some light on how the Sonos devices executes this.

KennethNielsen commented on tests/test_integration.py:
```diff
@@ -378,3 +378,362 @@ def test(self, soco):
+        test_pl = self._reset_spl_contents(soco)
+        num_tracks = soco.browse(ml_item=test_pl).total_matches
+        # get rid of the even numbered tracks
+        tracks = sorted([x for x in range(num_tracks) if not x & 1],
```
I'd prefer the form without bitwise operators for something like this: if x % 2 == 1

KennethNielsen commented on tests/test_integration.py:
```diff
@@ -378,3 +378,362 @@ def test(self, soco):
+        return test_pl
+
+    def test_reverse_track_order(self, soco):
+        """ Test reversing the tracks in the Sonos playlist. """
```
I think @lawrenceakka is in the process of correcting all the docstrings in another PR. I think the correct form is ```"""Test reversing the tracks in the Sonos playlist."""```. Followed by full stop, but no space before or after.

lawrenceakka replied:
Yup.

dundeemt replied:
Ok, just following what the prior code had done.

KennethNielsen replied:
@dundeemt Yeeess :$ not particularly consistent up to this, sorry that this bites you.

KennethNielsen commented:
Hallo @dundeemt 

First of all, really nice work. I especially enjoy the generous amount of integration tests \o/ (which really are functional tests, I will have to rename this soon).

I left some inline comments. Forgive the number of them, I tend to be a bit nit-picky. If you disagree with any of them, just let me know so we can discuss.

A few general thoughts came up while reviewing.

As you may have noticed, I was rambling on and on about whet the Sonos devices actually do, when you hand them a list of arguments. If it forms a single query to its internal database or executes them one by one. The more I read on, it seems as if you have actually concluded, that it executes them one by one. If I am correct in this assumption, then:

 1. I think it is all the more prudent to collect interable inputs into a single argument and always make just the one request, as that is what will be done on the other side anyway and it will have a (potentially large) performance impact.
 2. I think this one-by-one execution mode **and** its implications for deletions **and** the fact that it happens whether the user suppliers lists or strs should be clearly highlighted in the docstring, maybe with a Note or Warning box.

With regards to @lawrenceakka comments about using indexing. While I agree that it maybe could create a nicer interface, I still think we should also have the normal methods as implemented here. So that in the future, one would be able to to either or.

lawrenceakka commented on soco/core.py:
```diff
@@ -1255,6 +1266,35 @@ def get_sonos_playlists(self, *args, **kwargs):
         return self.get_music_library_information(*args, **kwargs)
 
+    def get_sonos_playlist_by_attr(self, attr_name, match):
+        """ Return the first Sonos Playlist DidlPlaylistContainer that
```
Please remove the space at the start!

lawrenceakka commented:
@dundeemt Sorry you have run into a barrage of comments. I think this is an important addition, and once we can get the few wrinkles ironed out, we will commit.  Please let us know if you have any questions, and feel free to argue back!

dundeemt commented:
@KennethNielsen @lawrenceakka  No need to argue back.  I understand your requests and will try and demonstrate the underlying calls unique rule set.

The following underlying calls return Work, for some value of work,  - No Error is returned
```python
[(u'InstanceID', 0), (u'ObjectID', u'SQ:143'), (u'UpdateID', '5'), (u'TrackList', u'3,2,1,0'), (u'NewPositionList', u'0,1,2,3')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:147'), (u'UpdateID', '5'), (u'TrackList', u'0'), (u'NewPositionList', u'3')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:149'), (u'UpdateID', '5'), (u'TrackList', u'2'), (u'NewPositionList', u'1')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:148'), (u'UpdateID', '5'), (u'TrackList', u'1'), (u'NewPositionList', u'2')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:150'), (u'UpdateID', '5'), (u'TrackList', u'3'), (u'NewPositionList', u'0')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:152'), (u'UpdateID', '5'), (u'TrackList', u'3,2'), (u'NewPositionList', u'0,1')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:153'), (u'UpdateID', '5'), (u'TrackList', u'3,1'), (u'NewPositionList', u'0,2')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:154'), (u'UpdateID', '5'), (u'TrackList', u'3,0'), (u'NewPositionList', u'0,3')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:156'), (u'UpdateID', '5'), (u'TrackList', u'3,1,0'), (u'NewPositionList', u'0,2')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:158'), (u'UpdateID', '5'), (u'TrackList', u'0,3,1'), (u'NewPositionList', u',0,2')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:159'), (u'UpdateID', '5'), (u'TrackList', u'3,0,1'), (u'NewPositionList', u'0,,2')]
```

However, the following calls return a UPNP 402 - Invalid Arguments
```python
[(u'InstanceID', 0), (u'ObjectID', u'SQ:144'), (u'UpdateID', '5'), (u'TrackList', u'0,1,2,3'), (u'NewPositionList', u'3,2,1,0')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:145'), (u'UpdateID', '5'), (u'TrackList', u'0,1,2'), (u'NewPositionList', u'3,2,1')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:146'), (u'UpdateID', '5'), (u'TrackList', u'0,1'), (u'NewPositionList', u'3,2')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:151'), (u'UpdateID', '5'), (u'TrackList', u'2,3'), (u'NewPositionList', u'1,0')]

[(u'InstanceID', 0), (u'ObjectID', u'SQ:155'), (u'UpdateID', '5'), (u'TrackList', u'3,0,1'), (u'NewPositionList', u'0,3,2')]
```
It appears, and I encourage you to test it/experiment yourself, that moves must be listed in reverse order while deletes can occur interspersed among moves.  The resulting changes appear as a single unit, but their implementation makes it apparent that this is an illusion.

To take any old list from a caller and attempt to do it in a single call to the underlying method would require that we "sort" their input, before relaying it to the underlying call.  Which is a recipe for trouble and bug reports.  I strongly urge that the helpers, in their simple form (a single int-like parameter for a move, remove) be kept in their simple form and that reorder_sonos_playlist be renamed _reorder_sonos_playlist to reinforce the notion, that you should know what your doing if you are going to use it.

I do not have windows at home so I have not used the desktop package, but the android version only allows you to do one reorder/remove operation at a time.  As for attempting to tie advance slicing operations to this, I have no interest in doing because I would never use it.

I am going to work on the docstrings, object over id, variable names and fixing the two bugs you have found.  I'll get that in place while everyone else has a chance to go over my findings and see if some one has an insight into solving the rules for multiple operations in one call.  Right now, clear_sonos_playlist is the only helper that takes advantage of it.

I am all good on changing the docstring formatting, but it would be good for them to be formalized in dev notes.  It is extremely hard to hit a moving target.  Look for an updated pull request later this weekend.    @KennethNielsen and you think you write long posts ; )

KennethNielsen commented:
> It appears, and I encourage you to test it/experiment yourself, that moves must be listed in reverse order while deletes can occur interspersed among moves. The resulting changes appear as a single unit, but their implementation makes it apparent that this is an illusion.

Oh my!

> To take any old list from a caller and attempt to do it in a single call to the underlying method would require that we "sort" their input, before relaying it to the underlying call. Which is a recipe for trouble and bug reports.

I agree completely!

>  I strongly urge that the helpers, in their simple form (a single int-like parameter for a move, remove) be kept in their simple form

Once again. I agree completely.

> and that reorder_sonos_playlist be renamed _reorder_sonos_playlist to reinforce the notion, that you should know what your doing if you are going to use it.

I don't think that is necessary. I like it that there is a more advanced method that gives you more power. If you simply add a warning to the docstring, along the lines of what you wrote above and your best guess at the rules of execution and saying that advanced users should themselves make sure that their calls does what they want them to, then I see no reason to make it pseudo private.

> I am going to work on the docstrings, object over id, variable names and fixing the two bugs you have found. I'll get that in place while everyone else has a chance to go over my findings and see if some one has an insight into solving the rules for multiple operations in one call. Right now, clear_sonos_playlist is the only helper that takes advantage of it.

I don't think I will have time to experiment myself, but as mentioned above, I'm am happy with the state in which clear_sonos_playlist is the only method that uses it, all other helpers and one by one and the advanced method has a verbose warning as outlined above.

> I am all good on changing the docstring formatting, but it would be good for them to be formalized in dev notes. It is extremely hard to hit a moving target.

Yes, definitely. The docs have for a long time been a weak spot and dev notes is a good idea. More than that, it is especially annoying when a convention is applied while you have an active PR. We apologize for that.

I look forward to the updated pull request. Please drop a comment when you are done.

dundeemt commented:
Ok, I've udpated doc strings, formatting, put object over id and I think the rest of the suggestions.  However this was a lengthy process, so please do not assume that if a change was requested and is not there that I disagreed, I most likely just missed it. So please bring it back up again.

dundeemt commented:
I messed up the merge some how and now I am in a hole.  I don't use git that much so its going to take a while to figure out.  Taking a break now.

dundeemt commented:
changed return type of reorder_sonos_playlist to a dict.  See Issue #348  Updated tests, refactored to remove deprecation warnings due to changes in project structure.

KennethNielsen commented on soco/core.py:
```diff
@@ -1578,6 +1580,246 @@ def _build_album_art_full_uri(self, url):
+                               update_id=0):
+        """Reorder and/or Remove tracks in a Sonos playlist.
+
+        The underlying call is quite complex as it both can move a track within
```
Shouldn't it be "as it can both"

KennethNielsen commented on soco/core.py:
```diff
@@ -1578,6 +1580,246 @@ def _build_album_art_full_uri(self, url):
+
+        If a list is specified for tracks, then a list must be used for
+        new_pos. Each list element is a discrete modification and the next
+        list operation must anticipate the new state of the playlist.
```
Very nice.

KennethNielsen commented on soco/core.py:
```diff
@@ -1578,6 +1580,246 @@ def _build_album_art_full_uri(self, url):
+        list operation must anticipate the new state of the playlist.
+
+        If a comma formatted string to tracks is specified, then use
+        a similiar string to specify new_pos. Those operations should be
```
I'm a little unsure what you mean here by "Those operations should be ordered from the end of the list to the beginning"

dundeemt replied:
Are you asking that I include an example?  i.e.
To reverse the order of the first 3 tracks, you should use  "2,0", "0,2"
 as compared to "0,2", "2,0".


On Mon, Sep 14, 2015 at 7:27 AM, Kenneth Nielsen <notifications@github.com>
wrote:

> In soco/core.py
> <https://github.com/SoCo/SoCo/pull/352#discussion_r39387395>:
>
> > @@ -1578,6 +1580,246 @@ def _build_album_art_full_uri(self, url):
> >      def _music_lib_search(self, search, start, max_items):
> >          return self.music_library._music_lib_search(search, start, max_items)
> >
> > +    @only_on_master
> > +    def reorder_sonos_playlist(self, sonos_playlist, tracks, new_pos,
> > +                               update_id=0):
> > +        """Reorder and/or Remove tracks in a Sonos playlist.
> > +
> > +        The underlying call is quite complex as it both can move a track within
> > +        the list or delete a track from the playlist.  All of this depends on
> > +        what tracks and new_pos specify.
> > +
> > +        If a list is specified for tracks, then a list must be used for
> > +        new_pos. Each list element is a discrete modification and the next
> > +        list operation must anticipate the new state of the playlist.
> > +
> > +        If a comma formatted string to tracks is specified, then use
> > +        a similiar string to specify new_pos. Those operations should be
>
> I'm a little unsure what you mean here by "Those operations should be
> ordered from the end of the list to the beginning"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/352/files#r39387395>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473

KennethNielsen commented on soco/core.py:
```diff
@@ -1578,6 +1580,246 @@ def _build_album_art_full_uri(self, url):
+            # sonos_playlist specified by the DidlPlaylistContainer object
+            sonos_playlist = device.get_sonos_playlists()[0]
+            device.reorder_sonos_playlist(sonos_playlist,
+                                          tracks=[0, ], new_pos=[1, ])
```
On the really nit-picky end of things, these commas in ``[0, ]`` are not necessary when dealing with lists the way they are for tuples. It is considered good style to add these trailing commas for **multi-line** lists (because it will make modifications easier to read in the version control diffs), but I believe not for single line lists.

dundeemt replied:
Commas make lists unambiguous.  While not required for a single integer,
['this',] != ['this'].  Any iterable will be split into elements within the
list if you don't mark them with commas.  If you would like me to remove
them, I will.  Personally, I don't like to vary based on data types, so I
choose to be explicit.   If you have a pep I'd be happy to read and perhaps
alter my opinion, but I've not read any thing authoritative against it.
 "Explicit is better than implicit." -
https://www.python.org/dev/peps/pep-0020/

On Mon, Sep 14, 2015 at 7:33 AM, Kenneth Nielsen <notifications@github.com>
wrote:

> In soco/core.py
> <https://github.com/SoCo/SoCo/pull/352#discussion_r39387778>:
>
> > +        ordered from the end of the list to the beginning
> > +
> > +        See the helper methods
> > +        :py:meth:`clear_sonos_playlist`, :py:meth:`move_in_sonos_playlist`,
> > +        :py:meth:`remove_from_sonos_playlist` for simplified usage.
> > +
> > +        update_id - If you have a series of operations, tracking the update_id
> > +        and setting it, will save a lookup operation.
> > +
> > +        Examples:
> > +          To reorder the first two tracks::
> > +
> > +            # sonos_playlist specified by the DidlPlaylistContainer object
> > +            sonos_playlist = device.get_sonos_playlists()[0]
> > +            device.reorder_sonos_playlist(sonos_playlist,
> > +                                          tracks=[0, ], new_pos=[1, ])
>
> On the really nit-picky end of things, these commas in [0, ] are not
> necessary when dealing with lists the way they are for tuples. It is
> considered good style to add these trailing commas for *multi-line* lists
> (because it will make modifications easier to read in the version control
> diffs), but I believe not for single line lists.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/352/files#r39387778>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473

KennethNielsen commented on soco/core.py:
```diff
@@ -1578,6 +1580,246 @@ def _build_album_art_full_uri(self, url):
+        Args:
+            sonos_playlist
+                (:py:class:`~.soco.data_structures.DidlPlaylistContainer`): The
+                Sonos playlist object or the item_id (str)of the Sonos
```
"(str)of" missing space

KennethNielsen commented on soco/core.py:
```diff
@@ -1578,6 +1580,246 @@ def _build_album_art_full_uri(self, url):
+                is done to find the correct value.
+
+        Returns:
+            dict: Which contains 3 elements. change, length and update_id.
```
"elements" colon, semi-colon?

dundeemt replied:
done

On Mon, Sep 14, 2015 at 7:36 AM, Kenneth Nielsen <notifications@github.com>
wrote:

> In soco/core.py
> <https://github.com/SoCo/SoCo/pull/352#discussion_r39387993>:
>
> > +                playlist.
> > +            tracks: (list): list of track indices(int) to reorder. May also be
> > +                a list of int like things. i.e. ``['0', '1',]`` OR it may be a
> > +                str of comma separated int like things. ``"0,1"``.  Tracks are
> > +                **0**-based. Meaning the first track is track 0, just like
> > +                indexing into a Python list.
> > +            new_pos (list): list of new positions (int|None)
> > +                corresponding to track_list. MUST be the same type as
> > +                ``tracks``. **0**-based, see tracks above. ``None`` is the
> > +                indicator to remove the track. If using a list of strings,
> > +                then a remove is indicated by an empty string.
> > +            update_id (int): operation id (default: 0) If set to 0, a lookup
> > +                is done to find the correct value.
> > +
> > +        Returns:
> > +            dict: Which contains 3 elements. change, length and update_id.
>
> "elements" colon, semi-colon?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/352/files#r39387993>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473

KennethNielsen commented on soco/core.py:
```diff
@@ -1578,6 +1580,246 @@ def _build_album_art_full_uri(self, url):
+
+        Example::
+
+            device.remove_from_sonos_playlist('SQ:0', track=0)
```
maybe ``sonos_playlist``here as well

dundeemt replied:
done

On Mon, Sep 14, 2015 at 7:41 AM, Kenneth Nielsen <notifications@github.com>
wrote:

> In soco/core.py
> <https://github.com/SoCo/SoCo/pull/352#discussion_r39388401>:
>
> > +            dict: See :py:meth:`reorder_sonos_playlist`
> > +
> > +        Raises:
> > +            SoCoUPnPException: See :py:meth:`reorder_sonos_playlist`
> > +        """
> > +        return self.reorder_sonos_playlist(sonos_playlist, int(track),
> > +                                           int(new_pos), update_id)
> > +
> > +    @only_on_master
> > +    def remove_from_sonos_playlist(self, sonos_playlist, track, update_id=0):
> > +        """Remove a track from a Sonos Playlist.
> > +        This is a convenience method for :py:meth:`reorder_sonos_playlist`.
> > +
> > +        Example::
> > +
> > +            device.remove_from_sonos_playlist('SQ:0', track=0)
>
> maybe sonos_playlisthere as well
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/352/files#r39388401>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473

KennethNielsen commented on soco/core.py:
```diff
@@ -1578,6 +1580,246 @@ def _build_album_art_full_uri(self, url):
+
+    @only_on_master
+    def get_sonos_playlist_by_attr(self, attr_name, match):
+        """ Return the first Sonos Playlist DidlPlaylistContainer that
```
No space before "Return"

dundeemt replied:
done

On Mon, Sep 14, 2015 at 7:43 AM, Kenneth Nielsen <notifications@github.com>
wrote:

> In soco/core.py
> <https://github.com/SoCo/SoCo/pull/352#discussion_r39388508>:
>
> > +                track is track 0, just like indexing into a Python list.
> > +            update_id (int): Optional update counter for the object. If left
> > +                at the default of 0, it will be looked up.
> > +
> > +        Returns:
> > +            dict: See :py:meth:`reorder_sonos_playlist`
> > +
> > +        Raises:
> > +            SoCoUPnPException: See :py:meth:`reorder_sonos_playlist`
> > +        """
> > +        return self.reorder_sonos_playlist(sonos_playlist, int(track), None,
> > +                                           update_id)
> > +
> > +    @only_on_master
> > +    def get_sonos_playlist_by_attr(self, attr_name, match):
> > +        """ Return the first Sonos Playlist DidlPlaylistContainer that
>
> No space before "Return"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/352/files#r39388508>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473

KennethNielsen commented:
@dundeemt Very nice work with the doc strings etc. I'm really happy with the result. I left just a hand full of inline comments on minor issues (typos etc.). Other than those, I consider this ready to merge. Just let us know when.

+1

dundeemt commented:
I'll merge and update.

dundeemt commented:
Once everyone is +1 with it, I can rebase and squash if you desire.

lawrenceakka commented:
@dundeemt please rebase, squash as appropriate, and your self to AUTHORS so I can commit

dundeemt commented:
working on the rebase now

On Sun, Sep 20, 2015 at 6:10 PM, Lawrence Akka <notifications@github.com>
wrote:

> @dundeemt <https://github.com/dundeemt> please rebase, squash as
> appropriate, and your self to AUTHORS so I can commit
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/352#issuecomment-141843855>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473

dundeemt commented:
working on squashing

lawrenceakka commented:
Thanks Jeff.  All done.  As a final task, please can you add a note to #332

dundeemt commented:
@KennethNielsen @lawrenceakka  Thank you for your help in getting this PR accepted and merged.  I appreciated your feedback, insights and efforts not only on the PR but the project in general.

KennethNielsen commented:
@dundeemt thank you for sticking with it. Sorry about not getting back to you sooner about some of the feedback, but I was pulled away by real life. The way I remember it, there was none of my comments in the last round that were really important, so I was happy to see that it had been pulled. Thanks for the effort. Just in the interest of discussion about style (which I must admit I lov), I'll comment on a single one of your comments from earlier.

KennethNielsen commented:
> Commas make lists unambiguous. While not required for a single integer, ['this',] != ['this']. Any iterable will be split into elements within the list if you don't mark them with commas. If you would like me to remove them, I will. Personally, I don't like to vary based on data types, so I choose to be explicit. If you have a pep I'd be happy to read and perhaps alter my opinion, but I've not read any thing authoritative against it. "Explicit is better than implicit." - https://www.python.org/dev/peps/pep-0020/

No. List already are unambiguous. If you try and type in your example, you will see that ``['this'] == ['this',]`` returns True. Now of course if you use the type function on a something that is iterable ``list('this')`` something else entirely happens. The way that I understand it, the trailing comma is without any semantic meaning whatsoever (except for tuples), but are **allowed**, and the main reason for that, as I understand it, is to make diffs look cleaner if you have multiline definitions. If you have a counter examples here, I'd be thrilled to see it. As far as tuples go, it is purely out of necessity, that a single element tuple has to be defined with a trailing comma, because the parenthesis is also used to function calls. So while, as you say, maybe it would be better to not have to vary depending on type, I really prefer to look at as if it is consistent without trailing comma for almost all container types (list, set, dict) and the tuple is just an ugly exception that we have to live with, but which is also easy to remember because we all know we use parenthesis for functions.

As far as ``"Explicit is better than implicit."``, I'm very aware of the Zen of Python and I think it, and the general focus on readability, is one of the absolutely brilliant things about the language. I'm just not sure I think it applies here, or that it overrules ``"Although practicality beats purity."``and most especially ``"Readability counts."`` ;)
